### DISCUSSION TITLE
Disable nvim-cmp autocomplete in Markdown files

### DISCUSSION THREAD

**Discussion by joeldrotleff** [NvChad/NvChad#2397] [Q&A] [answered]
I'd like to have autocomplete be disabled (so that the list of options doesn't pop up with useless suggestions when I'm just writing text.

Thanks,
Joel
![CleanShot 2023-09-23 at 19 19 29@2x](https://github.com/NvChad/NvChad/assets/2484624/8eeaec41-e63a-4a0b-b2b7-0000694ae2e6)

**Comment by joeldrotleff** [comment]
Hmm ok so I was able to get it working by changing init.lua to have this:

      require("cmp").setup(opts)

      require("cmp").setup.filetype('markdown', {
        sources = {},
      })

This works, but my understanding is ti breaks upgrading to make changes here, "All changes should be made in the /custom folder".

So I tried adding to overrides.lua:

M.cmp = {
   filetype = {
    "markdown", {
      sources = {}
    }
  }
}

But that doesn't work

**Reply by joachimverleysen** [reply]
> Hmm ok so I was able to get it working by changing init.lua to have this:
> 
> ```
>   require("cmp").setup.filetype('markdown', {
>     sources = {},
>   })
> ```

This is the one that worked for me, thanks

**Comment by siduck** [comment]
cmp has a enabled* function in its config, you can use that
  
```lua
  {
    "hrsh7th/nvim-cmp",
    opts = {
      enabled = function()
        return (vim.bo.ft ~= "markdown")
      end,
    },
  },
  ```
  
or you can make an autocmd that disables cmp on markdown filetypes

`require('cmp').setup.buffer { enabled = false }`

**Reply by joeldrotleff** [reply]
This worked great, thank you so much 🙏

**Reply by zwk5131** [reply]
This autocmd is amazing :D
``require('cmp').setup.buffer { enabled = false }``
How can i make a toggle function for this ?
@siduck

**Reply by siduck** [reply]
@zwk5131  read nvchad mapping docs

**Reply by zwk5131** [reply]
I did that before i asked... I'm asking because there wasn't a solution for a function where you can make a equals comparison if its enabled or not.. So my attempt is to do 2 Keymaps one to turn it on and one to turn it off... But that's not really nice...
Or maybe implement an if / else function where i can write the value true/false .. but i don't know how to implement in lua..

Only thing that I found is this:
```
-- For a more complex keymap
["<leader>tt"] = {
  function()
     require("base46").toggle_transparency()
  end,
  "toggle transparency",
},
```
But that didn't helped me...
https://nvchad.com/docs/config/mappings

So it would be nice if there is a solution or an example ... :)
 
@siduck

**Reply by siduck** [reply, marked as answer]
@zwk5131  just make a mapping with a function with your cmp code? 

something likethis

```lua
function()               
       vim.b.cmp_enabled = not vim.b.cmp_enabled
       require('cmp').setup.buffer { enabled = vim.b.cmp_enabled }
 end
 ```

**Reply by zwk5131** [reply]
Thank you for that good example :D
I came up with something like this :P

```
function()                 
        x = not x
	require('cmp').setup.buffer { enabled = not x }
end
```

I defined x globally maybe there is a nicer way :)

@siduck

**Reply by siduck** [reply]
@zwk5131 dont define it globally! , do vim.b , its buffer variable, if u really want it globally then do vim.g

**Reply by zwk5131** [reply]
okay, thanks :+1:

**Reply by kaelynhughes** [reply]
In case anyone else comes here when they're brand new to NvChad and confused by the file structure - I believe this goes in `/nvim/lua/plugins/init.lua`. At least, that's what worked for me!

**Comment by zwk5131** [comment]
Forgot to provide the solution for the toggle function... here it is:

Path:  `.config/nvim/lua/core` File: `mappings.lua`

M.general = {
  i = {

```
 -- Toggle Autocompletion for Writingmode
    ["<a-a>"] = { 
      function()                 
        vim.b.x = not vim.b.x
	require('cmp').setup.buffer { enabled = not vim.b.x }
    end,
      "Toggle Autocompletion CMP",
    },
```
},

THX again @siduck

**Reply by bjourne** [reply]
Should the path be `nvim/lua/mappings.lua` in recent NvChad versions?

**Comment by kharkwalsuraj** [comment]
### How to disable nvim-cmp by default and toggle it with <C-Space> in NvChad?
I'm using the latest version of NvChad, and I'm trying to disable nvim-cmp suggestions by default, and toggle them on/off using <C-Space>. Here's what I tried:

```lua
local map = vim.keymap.set

map({ 'n', 'i' }, '<C-Space>', function()
  vim.b.x = not vim.b.x
  require('cmp').setup.buffer { enabled = not vim.b.x }
end, { desc = "toggle suggestions" })
But it throws the following error:
```
### Output 
```
module 'cmp' not found:1
```
✅ What I want
nvim-cmp suggestions should be disabled by default

Pressing <C-Space> should toggle them on/off per buffer
